src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -750,5 +733,304 @@ private static string ToHtmlCase(string name)
+
+                var attributeName = _requiredAttributes.Substring(nameStartIndex, nameEndIndex - nameStartIndex);
+
+                return attributeName;
+            }
+
+            private TagHelperRequiredAttributeValueComparison ParseCssValueComparison(ErrorSink errorSink)
+            {
+                Debug.Assert(!AtEnd);
+                var valueComparison = TagHelperRequiredAttributeValueComparison.None;
+
+                if (CssValueComparisons.ContainsKey(Current))
+                {
+                    valueComparison = CssValueComparisons[Current];
+                    var op = Current;
+                    _index++;
+
+                    if (op != '=' && At('='))
+                    {
+                        // Two length operator (ex: ^=). Move past the second piece
+                        _index++;
+                    }
+                }
+
+                return valueComparison;
+            }
+
+            private string ParseCssValue(ErrorSink errorSink)
+            {
+                int valueStart;
+                int valueEnd;
+                if (At('\'') || At('"'))
+                {
+                    var quote = Current;
+
+                    // Move past the quote
+                    _index++;
+
+                    valueStart = _index;
+                    var quoteEndIndex = _requiredAttributes.IndexOf(quote, _index);
+                    if (quoteEndIndex == -1)
+                    {
+                        errorSink.OnError(
+                            SourceLocation.Zero,
+                            Resources.FormatTagHelperDescriptorFactory_InvalidRequiredAttributeMismatchedQuotes(
+                                _requiredAttributes,
+                                quote),
+                            length: 0);
+                        return null;
+                    }
+                    valueEnd = quoteEndIndex;
+                    _index = quoteEndIndex + 1;
+                }
+                else
+                {
+                    valueStart = _index;
+                    var valueEndIndex = _requiredAttributes.IndexOfAny(InvalidCssQuotelessValueCharacters, _index);
+                    valueEnd = valueEndIndex == -1 ? _requiredAttributes.Length : valueEndIndex;
+                    _index = valueEnd;
+                }
+
+                var value = _requiredAttributes.Substring(valueStart, valueEnd - valueStart);
+
+                return value;
+            }
+
+            private TagHelperRequiredAttributeDescriptor ParseCssSelector(ErrorSink errorSink)
+            {
+                Debug.Assert(At('['));
+
+                // Move past '['.
+                _index++;
+                PassOptionalWhitespace();
+
+                var attributeName = ParseCssAttributeName(errorSink);
+
+                if (attributeName == null ||
+                    !ValidateName(attributeName, targetingAttributes: true, errorSink: errorSink))
```

Comment:
Bit bizarre that `"["`, `"[   "` et cetera lead to an error message about the name length rather than the missing close square bracket.